docs_src/design/adr/device-service/0012-DeviceService-Filters.md

```diff
@@ -0,0 +1,152 @@
+- Two (Reading Filter) - after the sensor data has been collected and read and put into `Event/Reading` objects, there is a desire to filter some of the `Readings` based on the `Reading` values or `Reading` name (which is the device resource name and formerly the value descriptor name) or some combination of value and name.
+
+At this time, **this design only addresses the need for the second filter (Reading Filter)**.  At the time of this writing, no applicable use case has yet to be defined to warrant the Sensor Data Filter.
+
+### Reading Filters
+Reading filters will allow, not unlike application service filter functions today, to have `Readings` in an `Event` to be removed if:
+
+- the value was outside or inside some range, or the value was greater than, less than or equal to some value
+    - based on the `Reading` value (numeric) of a `Reading` outside a specified range (min/max) described in the device profile for a device resource.  Thus avoiding sending in outlier or jittery data `Readings` that could negatively effect analytics.
+    - based on the `Reading` value (numeric) equal to or near (with in some specified range) the last reading.  This allows a device service to reduce sending in `Event/Readings` that do not represent any significant change.  This differs from the already implemented onChangeOnly in that it is filtering `Readings` within a specified degree of change.
+- the value was the same as some or not the same as some specified value or values (for strings, boolean and other non-numeric values)
+- the name (the device resource name which used to be the value descriptor) matched a particular value; in other words match `temperature` or `humidity` as example device resources.
+
+Unlike application services, there is not a need to filter on a device name (or identifier).  Simply disable the device in the device service if all `Event/Readings` are to be stopped for the device.
+
+In the case that all `Readings` of an `Event` are filtered, it is assumed the entire `Event` is deemed to be worthless and not sent to core data by the device service.  If only some `Readings` from and `Event` are filtered, the `Event` minus the filtered `Readings` would be sent to core data.
+
+### Design / Architecture
+
+A new function interface shall be defined that, when implemented, performs a Reading Filter operation.  A ReadingFilter function would take a parameter (an `Event` containing readings), check whether the `Readings` of the `Event` match on the filtering configuration (see below) and if they do then remove them from the `Event`.  The ReadingFilter function would return the `Event` object (minus filtered `Readings`) or `nil` if the `Event` held no more `Readings`.  Pseudo code for the generic function is provided below.
+
+``` go
+func (f Filter) ReadingFilter(lc logger.LoggingClient, event *models.Event) (*models.Event, error) {
```

Comment:
See my previous comment about return codes... maybe this function should just return a bool indicating whether any readings had been removed due to filtering vs. returning a pointer to a new `Event`?